pkg/sql/opt/norm/general_funcs.go

```diff
@@ -1092,3 +1093,117 @@ func (c *CustomFuncs) DuplicateJoinPrivate(jp *memo.JoinPrivate) *memo.JoinPriva
 		SkipReorderJoins: jp.SkipReorderJoins,
 	}
 }
+
+// MakeKvScanFromZip takes a zip containing a call to
+// crdb_internal.scan and converts it into an
+// IndexScan expression, which can directly read from indexes.
+func (c *CustomFuncs) MakeKvScanFromZip(exp memo.RelExpr, zip memo.ZipExpr) memo.RelExpr {
+	item := zip.Child(0).(*memo.ZipItem)
+	scanExpr := item.Fn.(*memo.FunctionExpr)
+	var span tree.DArray
+	switch expr := scanExpr.Args.Child(0).(type) {
+	case *memo.FunctionExpr:
+		var tableIDVal opt.TableID
+		var indexIDVal cat.IndexOrdinal
+		if tableIDConstExpr, ok := expr.Args.Child(0).(*memo.ConstExpr); ok {
+			tableIDVal = opt.TableID(*tableIDConstExpr.Value.(*tree.DInt))
+		} else {
+			panic(errors.AssertionFailedf("index scans can only be generated from constants"))
+		}
+		if expr.Args.ChildCount() == 2 {
+			if indexIDConstExpr, ok := expr.Args.Child(1).(*memo.ConstExpr); ok {
+				indexIDVal = cat.IndexOrdinal(*indexIDConstExpr.Value.(*tree.DInt))
+			} else {
+				panic(errors.AssertionFailedf("index scans can only be generated from constants"))
+			}
+		}
+		var prefix roachpb.Key
+		if indexIDVal == 0 {
+			prefix = c.f.evalCtx.Codec.TablePrefix(uint32(tableIDVal))
+		} else {
+			prefix = c.f.evalCtx.Codec.IndexPrefix(uint32(tableIDVal), uint32(indexIDVal))
+		}
+
+		if err := span.Append(tree.NewDBytes(tree.DBytes(prefix))); err != nil {
+			panic(err)
+		}
+		if err := span.Append(tree.NewDBytes(tree.DBytes(prefix.PrefixEnd()))); err != nil {
+			panic(err)
+		}
+	case *memo.ConstExpr:
+		span = *expr.Value.(*tree.DArray)
+	}
+
+	// Collect metadata statistics if we are scanning a table
+	// and index with this query.
+	mdTableID, indexID := func() (opt.TableID, cat.IndexOrdinal) {
+		startKey := roachpb.Key(*span.Array[0].(*tree.DBytes))
+		_, tableID, err := c.f.evalCtx.Codec.DecodeTablePrefix(startKey)
+		if err != nil {
+			return 0, 0
+		}
+		// Error is intentionally ignored, if the index is missing.
+		_, _, indexID, _ := c.f.evalCtx.Codec.DecodeIndexPrefix(startKey)
+		ds, _, err := c.f.catalog.ResolveDataSourceByID(c.f.ctx, cat.Flags{AvoidDescriptorCaches: true}, cat.StableID(tableID))
+		if err != nil {
+			return 0, 0
+		}
```

Comment:
Should we be checking the endKey as well in case the span given crosses multiple table (which I think is possible right now if the user passes a constant span rather than using the helper functions?